dali/imgcodec/decoders/nvjpeg/memory_pool.h

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_IMGCODEC_DECODERS_NVJPEG_MEMORY_POOL_H_
+#define DALI_IMGCODEC_DECODERS_NVJPEG_MEMORY_POOL_H_
+
+#include <atomic>
+#include <thread>
+#include <map>
+#include <memory>
+#include <vector>
+#include "dali/core/mm/memory.h"
+#include "dali/core/mm/memory_kind.h"
+
+namespace dali {
+
+namespace imgcodec {
+
+namespace nvjpeg_memory {
+
+struct AllocInfo {
+  mm::memory_kind_id kind;
+  mm::Deleter deleter;
+  size_t size;
+  std::thread::id thread_id;
+};
+
+struct Deleter {
+  inline void operator()(void *p) const;
+};
+
+using unique_ptr_t = std::unique_ptr<char, Deleter>;
+
+struct Buffer {
+  Buffer(unique_ptr_t unq_ptr, mm::memory_kind_id kind, size_t sz);
+  unique_ptr_t ptr;
+  mm::memory_kind_id kind;
+  size_t size;
+};
+
+using MemoryPool = std::array<std::vector<Buffer>, static_cast<size_t>(mm::memory_kind_id::count)>;
+using BufferPool = std::map<std::thread::id, MemoryPool>;
+
+struct NVJpegMem {
```

Comment:
```suggestion
struct MemPool {
```
or similar

Reply:
It would affect NVJPEG decoder, though.

Reply:
`nvjpeg_memory.cc` is wrapping the calls, so it is not directly used in nvjpeg decoder.